File: libs/document-processor/src/registry/processors/myos-timeline-channel-processor.ts
import { BlueNode } from '@blue-labs/language';
import {
  TimelineEntrySchema,
  type TimelineEntry,
} from '@blue-repository/types/packages/conversation/schemas/TimelineEntry';
import {
  MyOSTimelineEntrySchema,
  type MyOSTimelineEntry,
} from '@blue-repository/types/packages/myos/schemas/MyOSTimelineEntry';
import { blueIds as myosBlueIds } from '@blue-repository/types/packages/myos/blue-ids';

import {
  myosTimelineChannelSchema,
  type MyOSTimelineChannel,
} from '../../model/index.js';
import type { ChannelEvaluationContext, ChannelProcessor } from '../types.js';
import { isTimelineEventNewer } from './shared/timeline-recency.js';

function resolveTimelineEntry(
  blue: ChannelEvaluationContext['blue'],
  event: BlueNode,
): MyOSTimelineEntry | TimelineEntry | null {
  if (blue.isTypeOf(event, MyOSTimelineEntrySchema)) {
    return blue.nodeToSchemaOutput(event, MyOSTimelineEntrySchema);
  }
  if (blue.isTypeOf(event, TimelineEntrySchema)) {
    return blue.nodeToSchemaOutput(event, TimelineEntrySchema);
  }
  return null;
}

export class MyOSTimelineChannelProcessor implements ChannelProcessor<MyOSTimelineChannel> {
  readonly kind = 'channel' as const;
  readonly blueIds = [myosBlueIds['MyOS/MyOS Timeline Channel']] as const;
  readonly schema = myosTimelineChannelSchema;

  matches(
    contract: MyOSTimelineChannel,
    context: ChannelEvaluationContext,
  ): boolean {
    const { event, blue } = context;
    if (!event) {
      return false;
    }

    const entry = resolveTimelineEntry(blue, event);
    if (!entry) {
      return false;
    }

    const entryTimelineId = entry.timeline?.timelineId;
    if (!entryTimelineId || !contract.timelineId) {
      return false;
    }

    return entryTimelineId === contract.timelineId;
  }

  channelize(
    contract: MyOSTimelineChannel,
    context: ChannelEvaluationContext,
  ): BlueNode | null | undefined {
    const { event, blue } = context;
    if (!event) {
      return null;
    }

    const entry = resolveTimelineEntry(blue, event);
    if (!entry) {
      return null;
    }

    return event.clone();
  }

  isNewerEvent(
    contract: MyOSTimelineChannel,
    context: ChannelEvaluationContext,
    lastEvent: BlueNode,
  ): boolean {
    const { event, blue } = context;
    if (!event) {
      return true;
    }
    return isTimelineEventNewer(blue, event, lastEvent);
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/registry/processors/MyOSTimelineChannelProcessor.java
package blue.language.processor.registry.processors;

import blue.language.model.Node;
import blue.language.processor.ChannelEvaluationContext;
import blue.language.processor.ChannelProcessor;
import blue.language.processor.model.MyOSTimelineChannel;
import blue.language.NodeProvider;

public class MyOSTimelineChannelProcessor implements ChannelProcessor<MyOSTimelineChannel> {

    @Override
    public Class<MyOSTimelineChannel> contractType() {
        return MyOSTimelineChannel.class;
    }

    @Override
    public boolean matches(MyOSTimelineChannel contract, ChannelEvaluationContext context) {
        if (!TimelineEventSupport.isConversationOrMyOSTimelineEntry(context.event())) {
            return false;
        }
        String expectedTimelineId = contract.getTimelineId();
        if (expectedTimelineId == null || expectedTimelineId.trim().isEmpty()) {
            return false;
        }
        String eventTimelineId = TimelineEventSupport.timelineId(context.event());
        if (!expectedTimelineId.trim().equals(eventTimelineId)) {
            return false;
        }
        NodeProvider nodeProvider = context.registry() != null ? context.registry().nodeProvider() : null;
        return WorkflowContractSupport.matchesEventFilter(context.event(), contract.getEvent(), nodeProvider);
    }

    @Override
    public Node channelize(MyOSTimelineChannel contract, ChannelEvaluationContext context) {
        Node event = context.event();
        if (!TimelineEventSupport.isConversationOrMyOSTimelineEntry(event)) {
            return null;
        }
        return event.clone();
    }

    @Override
    public String eventId(MyOSTimelineChannel contract, ChannelEvaluationContext context) {
        return TimelineEventSupport.eventId(context.event());
    }

    @Override
    public boolean isNewerEvent(MyOSTimelineChannel contract, ChannelEvaluationContext context, Node lastEvent) {
        return TimelineEventSupport.isNewer(context.event(), lastEvent);
    }
}
